### PR TITLE
network: Adds IPVLAN support

### PR DESCRIPTION
**Pull request by tomponline**
Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>

### CODE REVIEW

**Comment by tomponline** [comment]
jenkins: test this please

**Comment by tomponline** [comment]
@stgraber this is ready for review, tests seem fine (one rsync error which I think is unrelated).

**Comment by stgraber** [comment]
Small nit, your git config should be set so the Signed-off-by uses your Firstname and Lastname

File: doc/api-extensions.md [hunk 1]
```diff
@@ -757,3 +757,6 @@ migration is required.
 If the kernel supports seccomp-based syscall interception LXD can be notified
 by a container that a registered syscall has been performed. LXD can then
 decide to trigger various actions.
+
+## container\_nic\_ipvlan
```
Comment (stgraber) [inline doc/api-extensions.md]: This should be in a separate `api: IPVLAN support` commit along with adding the string to `shared/version/api.go`.

File: doc/containers.md [hunk 2]
```diff
@@ -249,8 +249,8 @@ hwaddr                  | string    | randomly assigned | no        | bridged, m
...
-ipv4.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv4 address to assign to the container through DHCP (bridged) and statically (ipvlan)
-ipv6.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv6 address to assign to the container through DHCP (bridged) and statically (ipvlan)
+ipv4.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv4 address to assign to the container through DHCP (bridged), for IPVLAN space delimited list of static addresses (at least 1 required)
+ipv6.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv6 address to assign to the container through DHCP (bridged), for IPVLAN space delimited list of static addresses (at least 1 required)
```
Comment (stgraber) [inline doc/containers.md]: Those should go in the doc commit

File: lxc/network.go [hunk 3]
```diff
@@ -157,6 +157,10 @@ func (c *cmdNetworkAttach) Run(cmd *cobra.Command, args []string) error {
 		device["nictype"] = "bridged"
 	}
 
+	if network.Type == "ipvlan" {
+		device["nictype"] = "ipvlan"
+	}
```
Comment (stgraber) [inline lxc/network.go]: Those changes to `lxc/network.go` aren't needed. The `network.Type` here refers to the parent host device as listed in `lxc network list`, there is no such thing as a `ipvlan` one of those.

File: lxc/network.go [hunk 4]
```diff
@@ -230,6 +234,10 @@ func (c *cmdNetworkAttachProfile) Run(cmd *cobra.Command, args []string) error {
 		device["nictype"] = "bridged"
 	}
 
+	if network.Type == "ipvlan" {
+		device["nictype"] = "ipvlan"
```
Comment (stgraber) [inline lxc/network.go]: Same

File: test/suites/container_devices.sh [hunk 5]
```diff
@@ -172,7 +172,51 @@ test_container_devices_nic() {
   # Test removing non-veth macvlan from running container.
   lxc config device remove "${ct_name}" eth1
 
-  # Cleanup.
+  # Cleanup veth/macvlan checks.
+  lxc delete "${ct_name}" -f
+  ip link delete "${ct_name}" type dummy
+
+  # Test ipvlan support to offline container (hot plugging not supported).
+  echo "Testing IPVLAN support..."
```
Comment (stgraber) [inline test/suites/container_devices.sh]: We don't usually do that

File: test/suites/container_devices.sh [hunk 6]
```diff
@@ -172,7 +172,51 @@ test_container_devices_nic() {
...
+  ip link add "${ct_name}" type dummy
+
+  # Check that starting IPVLAN container.
```
Comment (stgraber) [inline test/suites/container_devices.sh]: So this test is going to be problematic because if liblxc is lacking IPVLAN, this will all fail and Linux distributions which run our testsuite on their packages do not usually have the current master of liblxc.

I'm failing to come up with a cleaner way, so maybe have this deal properly with the error for missing lxc extension and move on?

File: doc/containers.md [hunk 7]
```diff
@@ -595,7 +599,7 @@ empty (default), no snapshots will be created. `snapshots.schedule.stopped`
 controls whether or not stopped container are to be automatically snapshotted.
 It defaults to `false`. `snapshots.pattern` takes a pongo2 template string,
 and the pongo2 context contains the `creation_date` variable. Be aware that you
-should format the date (e.g. use `{{ creation_date|date:"2006-01-02_15-04-05" }}`) 
+should format the date (e.g. use `{{ creation_date|date:"2006-01-02_15-04-05" }}`)
```
Comment (stgraber) [inline doc/containers.md]: Seems unrelated to this change and a trailing whitespace in markdown indicates a line break, so this may actually have made things worse :)

Comment (tomponline) [reply doc/containers.md]: @stgraber ah, looks like I included this accidentally after vscode auto removed the line, will take out

File: test/suites/container_devices.sh (hunk 6 above)
Comment (tomponline) [reply test/suites/container_devices.sh]: @stgraber how about adding a test go binary to the LXD test suite that loads liblxc using go-lxc and then allows us to query for supported LXC extensions, which we can then execute from the test suite?

File: test/suites/container_devices.sh (hunk 5 above)
Comment (tomponline) [reply test/suites/container_devices.sh]: @stgraber no problem, will remove.

File: lxc/network.go (hunk 3 above)
Comment (tomponline) [reply lxc/network.go]: @stgraber thanks, will take them out.

File: test/suites/container_devices.sh (hunk 6 above)
Comment (tomponline) [reply test/suites/container_devices.sh]: @stgraber actually, what about dynamically altering the output from `lxc info` in terms of the API extensions by modifying this file https://github.com/lxc/lxd/blob/master/shared/version/api.go#L23 to add the IPVLAN LXD functionality based on whether the LXC functionality is available (via go-lxc)? Then the tests can run `lxc info` to figure out whether they need to test that feature first.

Comment (stgraber) [reply test/suites/container_devices.sh]: Hmm, I was considering doing something similar for the seccomp one in liblxc but I think this will cause us problems. We should really keep things separate there.

We could possibly add a `driver_features` map similar to `kernel_features` to report support for those extensions we care about though.

Comment (tomponline) [reply test/suites/container_devices.sh]: @stgraber OK sounds good I will look at adding that.

Comment (tomponline) [reply test/suites/container_devices.sh]: @stgraber this has been added now. I've also split apart the tests for different nic types into different files as they were getting unwieldy.

**Comment by tomponline** [comment]
@stgraber ready for review again, thanks

File: doc/containers.md [hunk 8]
```diff
@@ -231,36 +231,37 @@ LXD supports different kind of network devices:
...
+mtu                     | integer   | parent MTU        | no        | all                                       | -                                      | The MTU of the new interface
+parent                  | string    | -                 | yes       | bridged, macvlan, ipvlan, physical, sriov | -                                      | The name of the host device or bridge
+vlan                    | integer   | -                 | no        | macvlan, ipvlan, physical                 | network\_vlan, network\_vlan\_physical | The VLAN ID to attach to
+ipv4.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv4 address to assign to the container through DHCP (bridged), for IPVLAN space delimited list of static addresses (at least 1 required)
+ipv6.address            | string    | -                 | no        | bridged, ipvlan                           | network                                | An IPv6 address to assign to the container through DHCP (bridged), for IPVLAN space delimited list of static addresses (at least 1 required)
```
Comment (stgraber) [inline doc/containers.md]: We usually prefer comma + optional space separated for such things.

File: doc/api-extensions.md [hunk 9]
```diff
@@ -760,3 +760,8 @@ decide to trigger various actions.
 
 ## container\_nic\_ipvlan
 This introduces the `ipvlan` "nic" device type.
+
+## lxc_features
```
Comment (stgraber) [inline doc/api-extensions.md]: The underscore needs escaping here

File: lxd/api_1.0.go [hunk 10]
```diff
@@ -211,6 +211,12 @@ func api10Get(d *Daemon, r *http.Request) Response {
 		"shiftfs":            fmt.Sprintf("%v", d.os.Shiftfs),
 	}
 
+	env.LXCFeatures = map[string]string{
```
Comment (stgraber) [inline lxd/api_1.0.go]: The various features in there should be checked at startup and cached as `lxc.HasApiExtension` is pretty costly and `/1.0` is our most queried API endpoint.

File: lxd/container_lxc.go [hunk 11]
```diff
@@ -8046,21 +8099,40 @@ func (c *containerLXC) insertNetworkDevice(name string, m types.Device) (types.D
...
 	}
 
 	return m, nil
 }
 
+// checkIPVLANSupport checks whether the liblxc available has the necessary IPVLAN features.
+func (c *containerLXC) checkIPVLANSupport() error {
+	if lxc.HasApiExtension("network_ipvlan") &&
+		lxc.HasApiExtension("network_l2proxy") &&
+		lxc.HasApiExtension("network_gateway_device_route") {
```
Comment (stgraber) [inline lxd/container_lxc.go]: With `lxc_features` moving this to the State struct, we can perform that check without actually hitting liblxc.

**Comment by stgraber** [comment]
Ok, so I'm going to take the `lxc_features` part of this and move it into its own PR as it needs to be split into separate commits too and we need to move the storage of those flags to the State struct with a startup time check to avoid constant jump to cgo.

We can then rebase this branch onto master once merged and just fix the IPv4 and IPv6 addresses list to be comma separated (with optional spaces).

**Comment by stgraber** [comment]
Did a rebase locally, switched to using OS.LXCFeatures and switched from space delimited to comma delimited with optional spaces.

Should be all good once the PR this now depends on are all merged.

**Comment by tomponline** [comment]
Thanks @stgraber, ill take onboard your comments about calling repeatedly into cgo being slow and the standard use of commas rather than spaces for option delimiters.

**Comment by s3rj1k** [comment]
@tomponline  In https://github.com/lxc/lxd/pull/5182 was discussion about GARP and GNDP, I did not see this implemented in this PR, not sure that closing that issue was good idea.

**Comment by stgraber** [comment]
@s3rj1k proxy ARP and proxy NDP are detected in the liblxc code and will fail with an error so long as not configured properly. Similar to what we discussed before, we will not have LXC or LXD alter those system wide settings.

**Comment by stgraber** [comment]
To clarify a bit, the liblxc ipvlan code does the needed sysctls on the generated interface itself but will not alter interfaces which are outside of liblxc's control, if those aren't set to the right value, an error is raised.

@tomponline is the error handling for this good enough for LXD or would we want to improve how we surface liblxc's errors on start or alternatively have a check for those sysctls directly in LXD?

**Comment by s3rj1k** [comment]
@stgraber ProxyARP and ProxyNDP is not the same as GARP and GNDP.
@tomponline  knows more, we had a discussion in the original PR about it.

@stgraber I am glad that IPVLAN finely is in the code base, but you are two quick to discard community efforts and issues. This makes community contribution to project almost impossible.

**Comment by tomponline** [comment]
@s3rj1k Yep I know what you mean about GARP I'll post a follow up shortly :)

**Comment by tomponline** [comment]
@stgraber the error handling could definitely be made clearer, currently if the required sysctls are missing then the container will not start, and you have to run the suggested lxd log retrieval command for the container to see the underlying LXC error. Its there, but takes some effort to get to.

Adding explicit checks for the same sysctls in LXD would certainly help the experience, so I can add a check for those no problem.

@s3rj1k the IPVLAN functionality has been implemented in the underlying LXC codebase so as to provide the functionality to a wider set of users. It was implemented as 3 discrete sets of functionality:

- The core IPVLAN device type - along with exposing the mode (l3, l3s and l2) and isolation flags (bridge, private, vepa).
- The layer 2 proxy (l2proxy) - this detects the sysctls required to allow Linux to generate proxy ARP/NDP responses (and generates errors if they are not set) and adds the necessary IP neigh proxy entries and static routes to allow Linux to generate the response packets. The intention with splitting this out into its own feature separate from IPVLAN itself is so that it can be used with a routed veth mode we are planning to add in the coming months.
- Device route gateway support - this allows the containers to use a device rather than an IP address as its default gateway, as using an IP for the gateway doesn't make sense for IPVLAN.

Then in LXD we made use of this new functionality to implement a simplified IPVLAN mode similar to your original pull-request and the existing MACVLAN experience in LXD.

Specifically:

- Only l3s mode with bridge isolation flag is currently supported in LXD, as this is required in order to work with the l2proxy mode (for some reason Linux will not generate proxy ARP/NDP packets if the IPVLAN device is operating in another mode). This is equivalent to the macvlan experience in LXD today.
- l2proxy mode is enabled, which assumes that the normal usage scenario for this mode is to add IPs to the containers from the local subnet of the parent interface (as opposed to other subnets being routed to the host). Not that this precludes you from doing that if you want though.

Additionally, the specific sysctls required have been slightly modified from your original PR, as during my testing I found that only these 2 are specifically required when using l3s mode and proxy ARP/NDP:

```
net.ipv6.conf.<parent>.forwarding=1
net.ipv6.conf.<parent>.proxy_ndp=1
```

I found that these sysctls from the original pull-request weren't specifically needed:

```
net.ipv4.ip_forward=1
net.ipv6.conf.all.forwarding=1
net.ipv4.conf.DEV.rp_filter=0
```

@s3rj1k the `rp_filter` one specifically I would be interested in knowing in which situations this one was needed?

Finally, @s3rj1k is correct that we did originally discuss adding gratuitous ARP and NDP adverts when a container boots so as to announce to the network which is slightly different to the proxy ARP/NDP mode in that the packets are sent out into the network irrespective of whether a device is asking "who has xxx?". This is primarily used for devices on the network that have already resolved the layer 2 address of the container and cached it in their IP neighbour cache. If the container is then migrated to a different host with a different MAC address then the devices on the network could continue to send packets to the old host as they had cached the MAC address (until such time as the cache expires or the OS notices there is no response and re-issues a "who has" request).

This could be added as an extension to the l2proxy mode in LXC (although the code I have today for it is written in Go rather than C so would need to be re-worked). This would then benefit both the IPVLAN mode (without code changes to LXD) and the future routed veth mode.

**Comment by s3rj1k** [comment]
@tomponline Enabling IPForwarding is a good idea because of inconsistent kernel behaviour, from what I observed older kernels need IP Forwarding enabled, newer ones enable this behaviour then IPVLAN L3s mode is enabled with proxy ARP/NDP records.

also per the kernel docs `net.ipv6.conf.all.forwarding=1` behaves differently from `net.ipv4.ip_forward=1`
for IPv4 it enables forwarding globally, for IPv6 it informs that you intend to forward traffic from specific interface (per interface forwarding).

Forwarding must be mentioned in documentation. But in my opinion this is needed in code to have defined behaviour across kernel versions.

rp_filter is needed for IPv4 then you have multiple Vlans on parent interface and inside CT.
with rp_filter disabled kernel will drop packets originated from different network.

As a minimum also must be in documentation.

I managed to get working GARP in my test environment, so this one is simple.

The NDP (IPv6) part is not so simple (does not work in my setup).

**Comment by lhorace** [comment]
@s3rj1k I agree with you re the community participation since I've been following this closely, nevertheless, it's implemented! So continue to contribute because it helps a lot of people too

**Comment by s3rj1k** [comment]
@lhorace Agreed, @tomponline did a great job.
Nevertheless guys from canonical should extend contributors documentation on how to write properly styled code.

Maybe add linters to project, I am willing to assist with this. @stgraber

**Comment by tomponline** [comment]
@s3rj1k RE the ipv4 forwarding sysctls, I made a mistake and missed out the `net.ipv4.conf.<parent>.forwarding=1` one which is checked for in LXC/LXD and is in the docs here: https://linuxcontainers.org/lxc/manpages//man5/lxc.container.conf.5.html (search for l2proxy).

I am going to add better error output to LXD so that it checks for those 3 sysctls directly and lets the user know if they are incorrect.

I've updated my original post here:
https://github.com/lxc/lxd/pull/5716#issuecomment-490811266

Regarding your note about older kernels requiring additional global forwarding sysctls enabled, I thought I would have a go at re-creating this using a CentOS 7 box which uses an older kernel 3.10.0-957.5.1.el7.x86_64 (it doesnt have IPVLAN support anyway, so its moot, but useful for testing the behaviour of proxy ARP/NDP settings).

For this experiment, the aim was to see if I could get a test host to response to ARP and NDP requests for an IP that it didn't have on its LAN interface, but did have a route to via the lo interface. As I am not testing "routed mode" proper (as that is not required for IPVLAN l3s mode), I just needed to check what the ping packets were arriving at the LAN interface of the test node to consider the test a success (as opposed to requiring a response to be generated) - as this will indicate that APR and NDP resolution has succeeded.

Here are the test steps I ran:

Disable proxy_arp and IPv4 forwarding globally so we have a baseline config:

```
sysctl net.ipv4.ip_forward=0
sysctl net.ipv4.conf.all.proxy_arp=0
sysctl net.ipv4.conf.all.forwarding=0
```

Enable forwarding on the LAN interface of the node:

```
sysctl net.ipv4.conf.enp2s0.forwarding=1
```

Add manual IP proxy entry on LAN interface (this is where IPv4 differs to IPv6 in the kernel because adding a manual proxy entry activates proxy ARP on that interface without needing to set it in the sysctls):

```
ip neigh add proxy 192.168.1.200 dev enp2s0
```

Add a route to the IP via the lo interface (this is to allow proxy ARP to work, using any non-LAN interface works the same):

```
ip r add 192.168.1.200/32 dev lo
```

Then run a tcpdump session:

```
tcpdump -i enp2s0 host 192.168.1.200 -nn
```

On another device connected to the same LAN, clear the local neighbour cache and ping 192.168.1.200

```
ip neigh delete 192.168.1.200 dev enp3s0
ping 192.168.1.200
```

We should expect to see this on TCPDUMP:

ARP resolution occurring, ICMP requests arriving at LAN interface (no ICMP reply expected though).

```
[root@rh-vz02 ~]# tcpdump -i enp2s0 host 192.168.1.200 -nn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp2s0, link-type EN10MB (Ethernet), capture size 262144 bytes
09:41:59.915189 ARP, Request who-has 192.168.1.200 tell 192.168.1.106, length 46
09:42:00.364437 ARP, Reply 192.168.1.200 is-at 28:92:4a:32:0c:e2, length 28
09:42:00.364557 IP 192.168.1.106 > 192.168.1.200: ICMP echo request, id 5567, seq 1, length 64
09:42:00.923362 IP 192.168.1.106 > 192.168.1.200: ICMP echo request, id 5567, seq 2, length 64
09:42:01.947367 IP 192.168.1.106 > 192.168.1.200: ICMP echo request, id 5567, seq 3, length 64
```

Similarly for IPv6:

```
sysctl net.ipv6.conf.all.forwarding=0
sysctl net.ipv6.conf.all.proxy_ndp=0
sysctl net.ipv6.conf.enp2s0.forwarding=1
sysctl net.ipv6.conf.enp2s0.proxy_ndp=1
```

```
ip -6 r add 2a02:xxx:76f4:1::200/128 dev lo
ip neigh add proxy 2a02:xxx:76f4:1::200 dev enp2s0
```

```
tcpdump -i enp2s0 host 2a02:xxx:76f4:1::200 -nn
```

On another device connected to the same LAN, clear the local neighbour cache and ping 2a02:xxx:76f4:1::200

```
ip neigh delete 2a02:xxx:76f4:1::200 dev enp3s0
ping6 2a02:xxx:76f4:1::200
```

We would expect to see this:

```
[root@rh-vz02 ~]# tcpdump -i enp2s0 host 2a02:xxx:76f4:1::200 -nn
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on enp2s0, link-type EN10MB (Ethernet), capture size 262144 bytes
10:14:27.398523 IP6 2a02:xxx:76f4:1:d553:a5ec:8306:c7ce > 2a02:xxx:76f4:1::200: ICMP6, echo request, seq 1, length 64
10:14:28.095726 IP6 2a02:xxx:76f4:1:d553:a5ec:8306:c7ce > 2a02:xxx:76f4:1::200: ICMP6, echo request, seq 2, length 64
10:14:29.116336 IP6 2a02:xxx:76f4:1:d553:a5ec:8306:c7ce > 2a02:xxx:76f4:1::200: ICMP6, echo request, seq 3, length 64
```

So as IPVLAN mode doesn't actually use the host as a router (it just needs to get the packets to arrive at the LAN interface and the IPVLAN device takes it from there), these settings seem to be the only ones required, even on older kernels without IPVLAN support.

I'd be interested to know about specific kernel versions that do not behave like this.

RE rp_filter, I'd be hesitant to encourage people to disable that feature in the docs, as it is intended to filter out spoofed packets, and, as I understand it, only should be disabled in situations where you have the LXD node multi-honed (as you say on multiple VLANs perhaps). Presumably this is true whether you are using ipvlan or some other sort of container networking (such as bridging) where packets arrive at a different interface than the replies would be delivered out of. Have I misunderstood? Is there something specific to ipvlan and rp_filter?

Thanks

**Comment by s3rj1k** [comment]
@tomponline You are correct about rp_filter usage, I am saying this should be documented case then multiple VLANS are needed.

For IPv4 IPForwarding per interface should be sufficient.
For IPv6 this is not the case, you need to enable `net.ipv6.conf.all.forwarding=1` as per documentation
https://github.com/torvalds/linux/blob/80f232121b69cc69a31ccb2b38c1665d770b0710/Documentation/networking/ip-sysctl.txt#L1519

To check this, use IPVLAN L2/L3 (without L3S) mode with IPv6

Actually L3 mode is in fact a router mode, see https://github.com/torvalds/linux/blob/6f0d349d922ba44e4348a17a78ea51b7135965b1/Documentation/networking/ipvlan.txt#L56

**Comment by tomponline** [comment]
@s3rj1k Yes the docs imply the host is performing routing, and in a way it is, but its more like an L3-switch, because the host does not appear in MTR traces as a traditional router hop, also even in l3s mode, the packets are filtered by the INPUT and OUTPUT chains in iptables rather than the FORWARD chain like a normally routed packet would.

I found this out to my disappointment when I tried to use ipvlan on my home router which has a direct ppp connection to the Internet independent of the LAN the ipvlan device was connected to. Packets always went out of the LAN interface rather than use the host's routing table.

With that in mind, I've re-tested proxy NDP with `net.ipv6.conf.all.forwarding` both on and off and my tests show the same result, that it can be disabled safely assuming you still have `net.ipv6.conf.enp3s0.forwarding=1` and it works for all modes, l2, l3 and l3s.

The problem comes when you bring IPv4 into the mix, as proxy ARP does not seem to work in l2 and l3 modes (and only l3s), even with `net.ipv4.conf.all.forwarding` and `net.ipv4.ip_forward=1` enabled.

This is why I settled on forcing l3s mode for LXD's IPVLAN.

**Comment by s3rj1k** [comment]
@tomponline actualy L2 mode is also usable, does not need ProxyARP or ProxyNDP, only IP Forwarding, look at output of `ip -d l` both parent and slave interfaces have ARP attribute. 

I would consider adding l2 mode as an option. Works similar to MacVtap but saves you MAC table entries on a dumb hardware switch.

IPVLAN is designed to send packets ONLY from parent interface, so no wonder you had issues with home router, to overcome this you need to set up policy routing.

Plain L3 mode is useless in my opinion.

**Comment by s3rj1k** [comment]
And I am actually a bit disappointed of @stgraber behaviour, did not even mentioned at all any work that was done by me, convincing him to actually do something about IPVLAN, spending lots of personal time rebasing, testing, rewriting initial PR.

May this be a warning to other members of community to think twice before contributing.

@tomponline thanks for your work.

**Comment by tomponline** [comment]
@s3rj1k OK thanks will give l2 another go, as you say l3 seems to be fairly redundant given l3s mode exists.

Now that the l3s mode that your original PR instigated has been merged, which put in place a lot of the underlying 'plumbing' into LXC (including l2 mode), adding an l2 mode, and required gateway params shouldn't be too hard. It may be that we use the presence of the "gateway" setting to switch into l2 mode. Although that has security implications (in that the host's firewall is bypassed on the inbound). I'll have a think and create a card to track the additional features we discussed here.

**Comment by tomponline** [comment]
@stgraber added better sysctl checks here https://github.com/lxc/lxd/pull/576
